design/one-pager-cross-resource-referencing.md

```diff
@@ -0,0 +1,343 @@
+to be provisioned. The YAML object will look like following:
+
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: VPC
+metadata:
+  namespace: cool-ns
+  name: my-vpc
+spec:
+  ...
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: Subnet
+metadata:
+  namespace: cool-ns
+  name: my-subnet
+spec:
+  # this is the vpcId of the external vpc, represented by my-vpc
+  vpcId: [my-vpc_vpcId]
+  ...
+```
+
+### Cross referencing using Reference Resolvers
+
+In this example since `vpcId` is non-deterministic we will need a mechanism to
+indicate this cross reference in the YAML object. I propose the notion of
+*Reference Resolver*, as a `go` interface as following:
+
+```go
+type ReferenceResolver interface {
+  // resolves the attribute, or returns an error
+  Resolve() (string,error)
+}
+```
+
+Having this interface, we can implement `VpcIDRefResolver` as
+
+```go
+type VpcIDRefResolver struct {
+  // the list of objects that are needed for resolving vpcID
+  // in this case we only need VPC name, but we could reference more objects
+  VPCName string `json:"vpcName"`
+}
+
+func (r *VpcIDRefResolver) Resolve()(string,error){
+  // implementation for building the attribute (in this case vpcID)
+  // from the referenced objects( in this case a VPC)
+}
+```
+
+Then we can modify [Subnet type] as:
+
+```diff
+- // VPCID is the ID of the VPC.
+- VPCID string `json:"vpcId"`
++ // VPCIDRef resolves the VPCID from the refenreced VPC
++ VPCIDRef VpcIDRefResolver `json:"vpcIdRef"`
+```
+
+Now we can update the `Subnet` YAML object in the sample solution as following:
+
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: Subnet
+metadata:
+  namespace: cool-ns
+  name: my-subnet
+spec:
+  # reference to API objects from which the vpcId will be resolved
+  vpcIdRef:
+    vpcName: my-vpc
+  ...
+---
+```
+
+Note here that we added a `Ref` suffix to the `vpcIdRef`, emphasizing that it is
+different than `vpcId`.
+
+This mechanism resolves the referenced non-deterministic attributes, as it waits
+for the specified object to become available. Until then, the `Resolve` method
+will return an error, postponing dependent resource reconciliation.
+
+For composite attributes, the corresponding reference resolver could specify all
+the required objects, and the function that produces the attribute. For instance
+consider the following solution in GCP:
+
+```yaml
+---
+apiVersion: compute.gcp.crossplane.io/v1alpha2
+kind: Network
+metadata:
+  name: my-network
+  namespace: cool-ns
+spec:
+  ...
+---
+apiVersion: compute.gcp.crossplane.io/v1alpha2
+kind: Subnetwork
+metadata:
+  name: my-subnetwork
+  namespace: cool-ns
+spec:
+  ...
+  network: /projects/myproject/global/networks/example
+  ...
+```
+
+To add `networkRef` field to the `Subnetwork`, we can implement
+`NetworkRefResolver` as:
+
+```go
+type NetworkRefResolver struct {
+  NetworkName string `json:"networkName"`
+  ProjectName string `json:"projectName"`
```

Comment:
I commented above, but I don't think we'd actually need (or want) to include `projectName` here. The name of the project would in all cases be the name of the project configured on the named network's GCP provider. I believe this pattern holds true for all composite references - that all information needed to form the actual reference are either direct or transitive properties of one API object.

Examples I can think of are:

* This example - project name is a discoverable transitive property of the referenced network.
* External resources that are unique within a zone or region (and thus referenced as something like `/regions/foo/resources/myresource` - region will a be direct property of the referenced resource.

Reply:
I understand that `projectName` could be retrieved as a transitive properties of one API object, but I tried to make an example for the generic case where multiple objects could be used to create an attribute. Would it be OK if I modify this after the first iteration of implementing the code?